subprojects/core/src/main/java/org/gradle/caching/internal/tasks/DefaultTaskOutputCachingBuildCacheKeyBuilder.java

```diff
@@ -90,11 +97,18 @@ public void appendTaskActionImplementations(Collection<ImplementationSnapshot> t
     }
 
     @Override
-    public void appendInputPropertyHash(String propertyName, HashCode hashCode) {
+    public void appendInputPropertyHash(String propertyName, @Nullable HashCode hashCode) {
         hasher.putString(propertyName);
-        hasher.putHash(hashCode);
-        inputHashes.put(propertyName, hashCode);
-        LOGGER.info("Appending inputPropertyHash for '{}' to build cache key: {}", propertyName, hashCode);
+        if (hashCode != null) {
+            hasher.putHash(hashCode);
+            inputHashes.put(propertyName, hashCode);
+            LOGGER.info("Appending inputPropertyHash for '{}' to build cache key: {}", propertyName, hashCode);
+        } else {
+            valid = false;
+            // Add a random hash to make sure it shows up as changed in the build scan
+            inputHashes.put(propertyName, HashCode.fromInt(new Random().nextInt()));
```

Comment:
I don't know about this. Are we sure it should show up as different? I think what we need is explicit support in build scans to mark invalid keys. Maybe what we need is to have `markInvalid()` to take a property name, so we could expose the property causing the invalidation towards build scans. But using a random number doesn't seem like the way we should handle this.

Reply:
I guess it should work. I changed the implementation to store the invalid property names separately.
I'll talk to the GE team if we should expose this list on the build operation, so the build scan plugin could consume it.

Reply:
:+1: